### PR TITLE
Persistant state for kademlia

### LINKED ISSUE #261: Persistent stateStore implementation for kademlia table

### PR DESCRIPTION
**Pull request by acud**
fixes #261

### CODE REVIEW

File: swarm/swarm.go [hunk 1]
```diff
@@ -149,15 +149,16 @@ func NewSwarm(ctx *node.ServiceContext, backend chequebook.Backend, config *api.
...
+	intervalsStore, err := state.NewDBStore(filepath.Join(config.Path, "stream-intervals.db"))
 	if err != nil {
 		return
 	}
 	self.streamer = stream.NewRegistry(addr, delivery, self.lstore, intervalsStore, false)
 	stream.RegisterSwarmSyncerServer(self.streamer, db)
 	stream.RegisterSwarmSyncerClient(self.streamer, db)
 
-	self.bzz = network.NewBzz(bzzconfig, to, nil)
+	stateStore, err := state.NewDBStore(filepath.Join(config.Path, "state-store.db"))
```
Comment (acud) [inline swarm/swarm.go]: should we use stateStore both for the kademlia and the `intervalsStore`? would make sense.. thoughts?

File: swarm/network/hive.go [hunk 2]
```diff
@@ -105,7 +105,7 @@ func (h *Hive) Start(server *p2p.Server) error {
 	log.Trace(fmt.Sprintf("%08x hive starting", h.BaseAddr()[:4]))
 	// if state store is specified, load peers to prepopulate the overlay address book
 	if h.Store != nil {
-		if err := h.loadPeers(); err != nil {
+		if err := h.loadPeers(); err != nil && err != state.ErrNotFound {
```
Comment (zelig) [inline swarm/network/hive.go]: put the error check in loadPeers

File: swarm/network/hive_test.go [hunk 3]
```diff
@@ -52,3 +59,50 @@ func TestRegisterAndConnect(t *testing.T) {
...
+	//start the hive and wait for the connection
+
+	pp.Start(s1.Server)
+	i := 0
+	pp.Overlay.EachAddr(nil, 256, func(addr OverlayAddr, po int, nn bool) bool {
+		delete(peers, addr.(*BzzAddr).String())
+		i++
+		return true
+	})
+	if len(peers) != 0 && i != 5 {
```
Comment (zelig) [inline swarm/network/hive_test.go]: this needs to be ||

File: swarm/state/dbstore_test.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+func testStore(t *testing.T, store Store) {
+	ser := &SerializingType{key: "key1", value: "value1"}
+
+	err := store.Put(ser.key, ser)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+}
+
```
Comment (zelig) [inline swarm/state/dbstore_test.go:100]: we should add test for another value like `[]string` that has no marshaller

File: swarm/state/memstore.go [hunk 5]
```diff
@@ -14,60 +14,62 @@
...
-		db: make(map[string]*Intervals),
+		db: make(map[string][]byte),
 	}
 }
 
 // Get retrieves Intervals for a specific key. If there is no Intervals
 // ErrNotFound is returned.
-func (s *MemStore) Get(key string) (i *Intervals, err error) {
+func (s *MemStore) Get(key string, i interface{}) (err error) {
+	_, ok := i.(encoding.BinaryUnmarshaler)
```
Comment (zelig) [inline swarm/state/memstore.go]: instead of `_` you assign the marshaler here and use it on line 54

File: swarm/state/memstore.go [hunk 6]
```diff
@@ -14,60 +14,62 @@
...
 	}
-	return i, nil
+
+	return i.(encoding.BinaryUnmarshaler).UnmarshalBinary(bytes)
 }
 
 // Put stores Intervals for a specific key.
-func (s *MemStore) Put(key string, i *Intervals) (err error) {
+func (s *MemStore) Put(key string, i interface{}) (err error) {
+	_, ok := i.(encoding.BinaryMarshaler)
```
Comment (zelig) [inline swarm/state/memstore.go]: same as above

File: swarm/swarm.go (hunk 1 above)
Comment (zelig) [reply swarm/swarm.go]: yes i would prefer

**Review by zelig — CHANGES_REQUESTED** [review]

File: swarm/state/memstore.go [hunk 7]
```diff
@@ -51,29 +47,36 @@ func (s *MemStore) Get(key string, i interface{}) (err error) {
...
 	}
 
-	bytes, err := i.(encoding.BinaryMarshaler).MarshalBinary()
-	if err != nil {
-		return err
-	}
+	return unmarshaler.UnmarshalBinary(bytes)
+}
 
+// Put stores Intervals for a specific key.
```
Comment (zelig) [inline swarm/state/memstore.go]: no intervals here, fix comments

**Review by zelig — CHANGES_REQUESTED** [review]

**Comment by slavikf** [comment]
N

**Review by zelig — APPROVED** [review]

**Pull request merged**